RecoTauTag/RecoTau/plugins/PFRecoTauChargedHadronProducer.cc

```diff
@@ -213,32 +213,7 @@ void PFRecoTauChargedHadronProducer::produce(edm::Event& evt, const edm::EventSe
       if (!(*outputSelector_)(*nextChargedHadron))
         continue;
 
-      const reco::Track* track = nullptr;
-      if (nextChargedHadron->getChargedPFCandidate().isNonnull()) {
-        const reco::PFCandidate* chargedPFCand =
-            dynamic_cast<const reco::PFCandidate*>(&*nextChargedHadron->getChargedPFCandidate());
-        if (chargedPFCand) {
-          if (chargedPFCand->trackRef().isNonnull())
-            track = chargedPFCand->trackRef().get();
-          else if (chargedPFCand->muonRef().isNonnull() && chargedPFCand->muonRef()->innerTrack().isNonnull())
-            track = chargedPFCand->muonRef()->innerTrack().get();
-          else if (chargedPFCand->muonRef().isNonnull() && chargedPFCand->muonRef()->globalTrack().isNonnull())
-            track = chargedPFCand->muonRef()->globalTrack().get();
-          else if (chargedPFCand->muonRef().isNonnull() && chargedPFCand->muonRef()->outerTrack().isNonnull())
-            track = chargedPFCand->muonRef()->outerTrack().get();
-          else if (chargedPFCand->gsfTrackRef().isNonnull())
-            track = chargedPFCand->gsfTrackRef().get();
-        } else {
-          track = nextChargedHadron->getChargedPFCandidate()->bestTrack();
-        }
-      }
-      if (track == nullptr) {
-        if (nextChargedHadron->getTrack().isNonnull()) {
-          track = nextChargedHadron->getTrack().get();
-        } else if (nextChargedHadron->getLostTrackCandidate().isNonnull()) {
-          track = nextChargedHadron->getLostTrackCandidate()->bestTrack();
-        }
-      }
+      const reco::Track* track = reco::tau::getTrackFromChargedHadron(*nextChargedHadron);
```

Comment:
Hi @mbluj , do I understand right that the `getTrack()` used in `reco::tau::getTrackFromChargedHadron` here:

https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoTauTag/RecoTau/src/pfRecoTauChargedHadronAuxFunctions.cc#L14-L16

return the `_track` private member of  `reco::PFRecoTauChargedHadron`, filled here:

https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoTauTag/RecoTau/plugins/PFRecoTauChargedHadronFromPFCandidatePlugin.cc#L189-L201

and the logic is the same as the one previously used in the `PFRecoTauChargedHadronProducer`?

Reply:
The `getTrack()` function from `cmssw/RecoTauTag/RecoTau/src/pfRecoTauChargedHadronAuxFunctions.cc` returns either private member `track_` of `reco::PFRecoTauChargedHadron` or if the instance of  `reco::PFRecoTauChargedHadron` is built with `pat::PackedCandidate`s (on top of miniAOD) it returns `bestTrack()` of of the charged candidate (track built from embedded information).
Logic of the 'getTrack()` function is not identical to the logic of replaced code as the sequence of options which are checked is a bit different, but the final result is identical. Moreover, the `getTrack()` function is used also in other places of tau reconstruction code, so its use also in this place make the code more coherent.

Reply:
Thanks for the explanation. Considering that the functionality was already implemented in `reco::tau::getTrackFromChargedHadron`, do you know why it was duplicated in `RecoTauTag/RecoTau/plugins/PFRecoTauChargedHadronProducer.cc`? Was there any specific reason?

Reply:
I think that there was not any specific reason, it was probably just forgotten when the code has been refactored in preparation for running with both reco::PFandidates (AOD mode) and pat::PackedCandidates (miniAOD mode).